chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -582,6 +542,92 @@ impl<'a, S: TransactionVerifierStorageRef, U: UtxosView> TransactionVerifier<'a,
         }
     }
 
+    pub fn connect_pos_accounting_outputs(
+        &mut self,
+        tx_source: TransactionSource,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
+        let tx_undo = tx
+            .outputs()
+            .iter()
+            .filter(|output| match output.purpose() {
+                OutputPurpose::StakePool(_) => true,
+                OutputPurpose::Transfer(_)
+                | OutputPurpose::LockThenTransfer(_, _)
+                | OutputPurpose::Burn => false,
+            })
+            .map(
+                |output| -> Result<PoSAccountingUndo, ConnectTransactionError> {
+                    match output.purpose() {
+                        OutputPurpose::StakePool(data) => {
+                            let tx_id = tx.get_id();
+                            let total_inputs = self.calculate_total_inputs(tx.inputs())?;
+                            ensure!(
+                                total_inputs.contains_key(&CoinOrTokenId::Coin),
+                                ConnectTransactionError::MissingCoinOutputToStake
+                            );
+                            ensure!(
+                                total_inputs.len() == 1,
+                                ConnectTransactionError::TokenInputForPoSAccountingOperation(tx_id)
+                            );
+                            // TODO: check StakePoolData fields
+
+                            let input0 = tx.inputs().get(0).expect("must be some");
+                            let delegation_amount = output.value().coin_amount().ok_or(
+                                ConnectTransactionError::TokenOutputForPoSAccountingOperation(
+                                    tx_id,
+                                ),
+                            )?;
+
+                            let (_, undo) = self
+                                .accounting_delta
+                                .create_pool(
+                                    input0.outpoint(),
+                                    delegation_amount,
+                                    data.decommission_key().clone(),
+                                )
+                                .map_err(ConnectTransactionError::PoSAccountingError)?;
+
+                            Ok(undo)
+                        }
+                        OutputPurpose::Transfer(_)
+                        | OutputPurpose::LockThenTransfer(_, _)
+                        | OutputPurpose::Burn => unreachable!(),
+                    }
+                },
```

Comment:
I need to propagate Err while mapping. Couldn't find better way to return Err and filter-map at the same time

Reply:
Hmm, that should be possible by returning from the `filter_map` closure:
* `None` if the item should be filtered out
* `Some(Err(e))` if propagating errors down the pipe
* `Some(Ok(blah))` in the success case

But maybe I missed something.

Reply:
This might work reasonably well:

Replace the `filter` with:
```rust
            .filter_map(|output| match output.purpose() {
                OutputPurpose::StakePool(data) => data,
                OutputPurpose::Transfer(_)
                | OutputPurpose::LockThenTransfer(_, _)
                | OutputPurpose::Burn => None,
            })
```

And in the subsequent `map`, you don't need to deal with the purpose matching and unreachables again

Reply:
you are right, thanks!